platform/cxxsupport/mstd_span

```diff
@@ -0,0 +1,367 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MSTD_SPAN_
+#define MSTD_SPAN_
+
+#include <array>
+#include <mstd_iterator>
+#include <mstd_type_traits>
+#include "mbed_assert.h"
+
+namespace mstd {
+
+constexpr size_t dynamic_extent = -1;
+
+template <typename, size_t = dynamic_extent> class span;
+
+namespace detail {
+
+template<typename ElementType, size_t Extent>
+class storage {
+public:
+    constexpr storage() noexcept = default;
+
+    constexpr storage(ElementType *ptr, size_t) noexcept :
+        _data(ptr)
+    {}
+
+    ElementType* _data = nullptr;
+    static constexpr size_t _size = Extent;
+};
+
+template<typename ElementType>
+class storage<ElementType, dynamic_extent> {
+public:
+    constexpr storage() noexcept = default;
+
+    constexpr storage(ElementType *ptr, size_t size) noexcept :
+        _data(ptr), _size(size)
+    {}
+
+    ElementType *_data = nullptr;
+    size_t _size = 0;
+};
+
+template<typename>
+struct is_span: std::false_type {};
+
+template<typename T, size_t E>
+struct is_span<span<T, E>>: std::true_type {};
+
+template<typename>
+struct is_std_array: std::false_type {};
+
+template<typename T, size_t N>
+struct is_std_array<std::array<T, N>>: std::true_type {};
+
+template<typename, typename = void>
+struct has_size : std::false_type {};
+
+template<typename T>
+struct has_size<T, void_t<decltype(mstd::size(std::declval<T>()))>>:
+    std::true_type {};
+
+template<typename, typename = void>
+struct has_data : std::false_type {};
+
+template<typename T>
+struct has_data<T, void_t<decltype(mstd::data(std::declval<T>()))>>:
+    std::true_type {};
+
+template<typename T, typename U = mstd::remove_cvref_t<T>>
+struct is_container{
+    static constexpr bool value =
+        not is_span<U>::value && not is_std_array<U>::value &&
+            not std::is_array<U>::value && has_size<T>::value &&
+                has_data<T>::value;
+};
+
+template <class T>
+using iterator_t = decltype(mstd::begin(std::declval<T&>()));
+
+template <class R>
+using range_reference_t = mstd::iter_reference_t<iterator_t<R>>;
+
+template <typename, typename, typename = void>
+struct is_compatible : std::false_type {};
+
+template <typename T, typename E>
+struct is_compatible<T, E,
+    typename mstd::enable_if_t<not mstd::is_same<
+            typename mstd::remove_cv_t<
+                decltype(mstd::data(std::declval<T>()))
+            >, void>::value>>:
+mstd::is_convertible<remove_pointer_t<
+    decltype(mstd::data(std::declval<T>()))> (*)[], E (*)[]>{};
+
+} // namespace detail
+
+template<typename ElementType, size_t Extent>
+class span {
```

Comment:
The model in other straightforward backport classes in cxxsupport is that we defer to the toolchain if it has the class available - use the feature test macros to see if `std::span` exists, and if so, include its header and use `using` to make `mstd::span` == `std:span`.

Although that's a bit trickier with a whole new header rather than an extension to an existing one - you probably want something like "if compiler is C++17 or later, and __has_include <span>, and after including that it has a new enough __cpp_lib_span macro."

We assume toolchain implementations will be better generally, and it gives you the necessary type compatibility for any other library calls taking the `std` type.

You probably can't test this now, but it's got a 90% chance of working when the C++20 toolchains comes along, and if it doesn't and breaks something, we or our descendents will find out then and be prompted to fix it.

Reply:
Implemented it and tested this on godbolt with GCC and it seems to work with C++20 so we should be fine for now.